src/Microsoft.Diagnostics.DebugServices.Implementation/ImageMappingMemoryService.cs

```diff
@@ -15,30 +15,46 @@ namespace Microsoft.Diagnostics.DebugServices.Implementation
     /// <summary>
     /// Memory service wrapper that maps and fixes up PE module on read memory errors.
     /// </summary>
-    internal class ImageMappingMemoryService : IMemoryService
+    public class ImageMappingMemoryService : IMemoryService, IDisposable
     {
+        private readonly IServiceContainer _container;
         private readonly IMemoryService _memoryService;
         private readonly IModuleService _moduleService;
         private readonly MemoryCache _memoryCache;
         private readonly HashSet<ulong> _recursionProtection;
 
         /// <summary>
-        /// The PE and ELF image mapping memory service. This service assumes that the managed PE 
-        /// assemblies are in the module service's list. This is true for dbgeng, dotnet-dump but not
-        /// true for lldb (only native modules are provided).
+        /// The PE, ELF and MacOS image mapping memory service. For the dotnet-dump linux dump reader and
+        /// dbgeng the native module service providers managed and modules, but under lldb only native 
+        /// modules are provided. The "managed" flag is for those later cases.
         /// </summary>
-        /// <param name="target">target instance</param>
+        /// <param name="container">service container</param>
         /// <param name="memoryService">memory service to wrap</param>
-        internal ImageMappingMemoryService(ITarget target, IMemoryService memoryService)
+        /// <param name="managed">if true, map managed modules, else native</param>
+        public ImageMappingMemoryService(IServiceContainer container, IMemoryService memoryService, bool managed)
         {
+            _container = container.Clone();
+            _container.AddService(memoryService);
             _memoryService = memoryService;
-            _moduleService = target.Services.GetService<IModuleService>();
+            _moduleService = managed ? new ManagedImageMappingModuleService(_container.Services) : _container.Services.GetService<IModuleService>();
             _memoryCache = new MemoryCache(ReadMemoryFromModule);
             _recursionProtection = new HashSet<ulong>();
-            target.OnFlushEvent.Register(_memoryCache.FlushCache);
-            target.DisposeOnDestroy(target.Services.GetService<ISymbolService>()?.OnChangeEvent.Register(_memoryCache.FlushCache));
+
+            ITarget target = _container.Services.GetService<ITarget>();
+            target.OnFlushEvent.Register(Flush);
+
+            IDisposable onChangeEvent = _container.Services.GetService<ISymbolService>()?.OnChangeEvent.Register(Flush);
```

Comment:
Should this always flush the memoryService if a symbol server got added?

Reply:
The _memoryCache could contain failures that are cached that might be resolved with a new symbol server.